vault/src/metrics.rs

```diff
@@ -0,0 +1,172 @@
+        Ok(v) => v,
+        Err(e) => {
+            eprintln!("metrics could not be parsed `from_utf8`: {}", e);
+            String::default()
+        }
+    };
+    buffer.clear();
+    res
+}
+
+pub async fn metrics_handler() -> Result<impl Reply, Rejection> {
+    let mut metrics = serialize(&REGISTRY.gather());
+    let custom_metrics = serialize(&gather());
+    metrics.push_str(&custom_metrics);
+    Ok(metrics)
+}
+
+pub async fn update_bridge_metrics(parachain_rpc: InterBtcParachain, vault_id: VaultId) -> Result<(), Error> {
+    let decimals_offset = FixedU128::one()
+        .into_inner()
+        .checked_div(vault_id.collateral_currency().one())
+        .unwrap_or_default() as f64;
+
+    let actual_collateral = parachain_rpc.get_vault_total_collateral(vault_id.clone()).await?;
+    let float_actual_collateral = FixedU128::from_inner(actual_collateral).to_float() * decimals_offset;
+
+    LOCKED_COLLATERAL.set(float_actual_collateral);
+
+    let collateralization = parachain_rpc
+        .get_collateralization_from_vault(vault_id.clone(), false)
+        // if the collateralization is infinite, return 0 rather than logging an error, so
+        // the metrics do change in case of a replacement
+        .await
+        .unwrap_or(0u128);
+    let float_collateralization_percentage = FixedU128::from_inner(collateralization).to_float();
+    COLLATERALIZATION.set(float_collateralization_percentage);
+
+    let required_collateral = parachain_rpc
+        .get_required_collateral_for_vault(vault_id.clone())
+        .await?;
+    let truncated_required_collateral = FixedU128::from_inner(required_collateral).to_float() * decimals_offset;
+    REQUIRED_COLLATERAL.set(truncated_required_collateral);
+    Ok(())
+}
+
+pub async fn update_bitcoin_metrics<B: BitcoinCoreApi + Clone + Send + Sync + 'static>(
+    btc_rpc: B,
+    btc_confirmations: u32,
+) {
+    match btc_rpc.get_balance(Some(btc_confirmations)).await {
+        Ok(bitcoin_balance) => LOCKED_BTC.set(bitcoin_balance.as_btc()),
+        Err(e) => {
+            // failed to cancel; get up-to-date request list in next iteration
+            tracing::error!("Failed to get Bitcoin balance: {}", e);
+        }
+    }
+    update_bitcoin_fees(btc_rpc, None).await;
+}
+
+pub async fn update_bitcoin_fees<B: BitcoinCoreApi + Clone + Send + Sync + 'static>(
+    btc_rpc: B,
+    new_fee_entry: Option<SignedAmount>,
+) {
+    let mut bitcoin_fees_vec = BITCOIN_FEES.write().await;
+    if bitcoin_fees_vec.len() == 0 && new_fee_entry.is_none() {
+        // Initialize the `AVERAGE_BTC_FEE` by fetching all Bitcoin txs.
+        match btc_rpc.list_transactions(None).await {
+            Ok(tx_list) => {
+                let mut fees: Vec<_> = tx_list
+                    .iter()
+                    .filter_map(|tx| tx.detail.fee.map(|amount| amount.as_btc()))
+                    .collect();
+
+                bitcoin_fees_vec.append(&mut fees);
```

Comment:
Unless you plan to use this vec for other statistics, there is no need no maintain this vec - just maintain two variables containing the total amount, and the number of transactions.